### DISCUSSION TITLE
[mysqlrouter_exporter ERROR] response status code is not 200 got 401

### DISCUSSION THREAD

**Discussion by pdpxy** [rluisr/mysqlrouter_exporter#27] [Q&A] [answered]
Hi,

Apologies if I'm missing something obvious here but I seem to be getting this error whenever I try to run:
`mysqlrouter_exporter --user=exporteruser --pass=12345678 --skip-tls-verify`

I have the user with credentials exporteruser/12345678 created for mysqlrouter. But somehow, in the logs it says **401**.
I'm using **mysql/mysql-router:8.0.29** as image

Inside the mysql-router container, I can verify that the user I created is there.
sh-4.4$ mysqlrouter_passwd list /opt/myrouter/mysqlrouter.pwd 
exporteruser:$5$jdCszkxyLmdUMGpr$75ketDPkNmwxLozb8bQE8M40iFvoiWc5Wwo7bTRy08C

**Comment by rluisr** [comment]
Could you paste mysqlrouter config?

**Reply by pdpxy** [reply]
`# File automatically generated during MySQL Router bootstrap
[DEFAULT]
name=plng_cabo_stg
logging_folder=/opt/myrouter/log
runtime_folder=/opt/myrouter/run
data_folder=/opt/myrouter/data
keyring_path=/opt/myrouter/data/keyring
master_key_path=/opt/myrouter/mysqlrouter.key
connect_timeout=5
read_timeout=30
dynamic_state=/opt/myrouter/data/state.json
client_ssl_cert=/opt/myrouter/data/router-cert.pem
client_ssl_key=/opt/myrouter/data/router-key.pem
client_ssl_mode=PREFERRED
server_ssl_mode=AS_CLIENT
server_ssl_verify=DISABLED
unknown_config_option=error

[logger]
level=DEBUG

[metadata_cache:bootstrap]
cluster_type=gr
router_id=138
user=clumon
metadata_cluster=plng_cabo_stg
ttl=60
auth_cache_ttl=-1
auth_cache_refresh_interval=60
use_gr_notifications=1

[routing:bootstrap_rw]
bind_address=0.0.0.0
bind_port=3306
destinations=metadata-cache://plng_cabo_stg/?role=PRIMARY
routing_strategy=first-available
protocol=classic

[routing:bootstrap_ro]
bind_address=0.0.0.0
bind_port=3307
destinations=metadata-cache://plng_cabo_stg/?role=SECONDARY
routing_strategy=round-robin-with-fallback
protocol=classic

[routing:bootstrap_x_rw]
bind_address=0.0.0.0
bind_port=3308
destinations=metadata-cache://plng_cabo_stg/?role=PRIMARY
routing_strategy=first-available
protocol=x

[routing:bootstrap_x_ro]
bind_address=0.0.0.0
bind_port=3309
destinations=metadata-cache://plng_cabo_stg/?role=SECONDARY
routing_strategy=round-robin-with-fallback
protocol=x

[http_server]
port=8443
ssl=1
ssl_cert=/opt/myrouter/data/router-cert.pem
ssl_key=/opt/myrouter/data/router-key.pem

[http_auth_realm:default_auth_realm]
backend=default_auth_backend
method=basic
name=default_realm

[rest_router]
require_realm=default_auth_realm

[rest_api]

[http_auth_backend:default_auth_backend]
backend=file
filename=/opt/myrouter/mysqlrouter.pwd

[rest_routing]
require_realm=default_auth_realm

[rest_metadata_cache]
require_realm=default_auth_realm`

**Reply by pdpxy** [reply]
I am creating the user as soon as the pod comes up.

echo -n '12345678' | mysqlrouter_passwd set /opt/myrouter/mysqlrouter.pwd exporteruser

**Reply by rluisr** [reply]
What passed `--url` in mysqlrouter_exporter?

> I am creating the user as soon as the pod comes up.

It means, creating a `mysqlrouter.pwd` file **before** running MySQL Router process right?

**Reply by pdpxy** [reply]
> What passed `--url` in mysqlrouter_exporter?
> 
> > I am creating the user as soon as the pod comes up.
> 
> It means, creating a `mysqlrouter.pwd` file **before** running MySQL Router process right?

MYSQLROUTER_EXPORTER_URL is "https://mysql-router-db03-internal-svc:8443" which is an internal service inside the kubernetes cluster. Port 8443 is the API port of the mysql router

Here's the command:
`echo -n '12345678' | mysqlrouter_passwd set /opt/myrouter/mysqlrouter.pwd exporteruser ; \n mysqlrouter \n --bootstrap=mysqlrouter@mydb-server:3306 \n --name=plng_cabo_stg \n --directory=/opt/myrouter \n --conf-use-gr-notifications \n --account=clumon \n --account-create=if-not-exists \n --conf-base-port=3306 \n --conf-set-option=logger.level=DEBUG \n --conf-set-option=http_auth_backend:default_auth_backend.backend=file \n --conf-set-option=http_auth_backend:default_auth_backend.filename=/opt/myrouter/mysqlrouter.pwd \n --force < /var/secrets/mysqlrouter.pass ; \n /opt/myrouter/stop.sh ; \n /opt/myrouter/start.sh"`

**Reply by rluisr** [reply]
> MYSQLROUTER_EXPORTER_URL is "https://mysql-router-db03-internal-svc:8443/" which is an internal service inside the kubernetes cluster. Port 8443 is the API port of the mysql router

It seems good.
mysqlrouter_exporter can't authenticate your MySQL Router BASIC authentication from your log: `response status code is not 200 got 401` 

But I have never deployed mysqlrouter_exporter and MySQL Router to Kubernetes so I cannot advise further...

Please try to build a container image including the pwd file and the conf file in the container image for **testing**.

**Reply by pdpxy** [reply]
I'm using mysql/mysql-router:8.0.29 image for mysql-router and this version of your exporter -- https://github.com/Marfeel/PubDockerFiles/blob/master/mysqlrouter-exporter/Dockerfile 

Are they compatible?

**Reply by pdpxy** [reply, marked as answer]
Hi @rluisr  - I have managed to make it work :)
What I did was basically restart the router after the bootstrap process and explicitly passing the conf file that was generated after creating the user and the bootstrapping. 

Some additional questions though:
1. Where can I find the documentation of the metrics that will become available once enabling the exporter?
2. Are these the only metrics available?
- mysqlrouter_router_status
- mysqlrouter_route_health
- mysqlrouter_route_destinations
- mysqlrouter_route_active_connections
- mysqlrouter_route_total_connections

I used the following in the args when running the exporter:
`--collect.metadata.status
              --collect.route.connections.byte_from_server
              --collect.route.connections.byte_to_server
              --collect.route.connections.time_started
              --collect.route.connections.time_connected_to_server
              --collect.route.connections.time_last_sent_to_server
              --collect.route.connections.time_received_from_server`

**Comment by rluisr** [comment]
Thank you for sharing that!

If you passed `--collect.metadata.status` you should be able to see `mysqlrouter_metadata*`

You can see gauges from https://github.com/rluisr/mysqlrouter_exporter/blob/master/gauges.go

**Reply by pdpxy** [reply]
Thank you!! Is there a metric that will ultimately say who is the primary member/leader in the DB cluster?

**Reply by rluisr** [reply]
You can see `mysqlrouter_route_destinations` but these are not gauges so you have to use `address` label and maybe MySQL Router API does not export about a member/leader.

In my environment, output is here:
```
# HELP mysqlrouter_route_destinations 
# TYPE mysqlrouter_route_destinations gauge
mysqlrouter_route_destinations{address="mysql11.luis.local",name="bootstrap_ro",port="3306"} 0
mysqlrouter_route_destinations{address="mysql11.luis.local",name="bootstrap_x_ro",port="33060"} 0
mysqlrouter_route_destinations{address="mysql12.luis.local",name="bootstrap_ro",port="3306"} 0
mysqlrouter_route_destinations{address="mysql12.luis.local",name="bootstrap_x_ro",port="33060"} 0
mysqlrouter_route_destinations{address="mysql13.luis.local",name="bootstrap_rw",port="3306"} 0
mysqlrouter_route_destinations{address="mysql13.luis.local",name="bootstrap_x_rw",port="33060"} 0
mysqlrouter_route_destinations{address="mysql15.luis.local",name="bootstrap_ro",port="3306"} 0
mysqlrouter_route_destinations{address="mysql15.luis.local",name="bootstrap_x_ro",port="33060"} 0
```
